File: src/main/java/arcana/client/model/wand/WandModelGeometry.java
package arcana.client.model.wand;

import arcana.common.items.wand.CapItem;
import arcana.common.items.wand.CoreItem;
import arcana.common.items.wand.WandItem;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import com.mojang.datafixers.util.Either;
import com.mojang.datafixers.util.Pair;
import net.minecraft.client.renderer.model.*;
import net.minecraft.client.renderer.texture.AtlasTexture;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.entity.LivingEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.vector.TransformationMatrix;
import net.minecraftforge.client.model.IModelConfiguration;
import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.client.model.ModelTransformComposition;
import net.minecraftforge.client.model.PerspectiveMapWrapper;
import net.minecraftforge.client.model.geometry.IModelGeometry;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Collection;
import java.util.Collections;
import java.util.Random;
import java.util.Set;
import java.util.function.Function;

import static arcana.utils.Util.arcLoc;
import static arcana.utils.Util.withPath;

@SuppressWarnings({"UnstableApiUsage", "deprecation"})
public class WandModelGeometry implements IModelGeometry<WandModelGeometry> {
    // hold onto data here
    ResourceLocation cap;
    ResourceLocation core;
    ResourceLocation variant;
    ResourceLocation focus;

    Logger LOGGER = LogManager.getLogger();

    public WandModelGeometry(ResourceLocation cap, ResourceLocation core, ResourceLocation variant, ResourceLocation focus) {
        this.cap = cap;
        this.core = core;
        this.variant = variant;
        this.focus = focus;
    }

    public IBakedModel bake(IModelConfiguration owner, ModelBakery bakery, Function<RenderMaterial, TextureAtlasSprite> spriteGetter, IModelTransform modelTransform, ItemOverrideList overrides, ResourceLocation modelLocation) {
        IModelTransform transformsFromModel = owner.getCombinedTransform();
        ImmutableList.Builder<BakedQuad> builder = ImmutableList.builder();
        ImmutableMap<ItemCameraTransforms.TransformType, TransformationMatrix> transformMap = PerspectiveMapWrapper.getTransforms(new ModelTransformComposition(transformsFromModel, modelTransform));

        // get core texture
        RenderMaterial coreTex = new RenderMaterial(AtlasTexture.LOCATION_BLOCKS, core);
        // get cap texture
        RenderMaterial capTex = new RenderMaterial(AtlasTexture.LOCATION_BLOCKS, cap);

        if(variant == null)
            variant = arcLoc("wand");

        ResourceLocation coreLoc = new ResourceLocation(variant.getNamespace(), "item/wands/variants/" + variant.getPath());
        //ResourceLocation coreLoc = arcLoc("item/wands/variants/wand");
        IUnbakedModel coreModel = bakery.getModel(coreLoc);
        ItemCameraTransforms tfs = ItemCameraTransforms.NO_TRANSFORMS;

        // they *should* be, but might as well check.
        if (coreModel instanceof BlockModel) {
            BlockModel model = (BlockModel) coreModel;
            model.textureMap.put("core", Either.left(coreTex));
            model.textureMap.put("cap", Either.left(capTex));
            tfs = model.getTransforms();
        } else {
            LOGGER.error("Wand model isn't a block model!");
        }

        // get focus model and texture, apply, and add
        Random rand = new Random();
        if (focus != null) {
                IBakedModel focusModel = bakery.getBakedModel(arcLoc("item/wands/foci/wand_focus"), modelTransform, spriteGetter);
            if(variant.getPath().equals("staff")) focusModel = bakery.getBakedModel(arcLoc("item/wands/foci/staff_focus"), modelTransform, spriteGetter);
            if (focusModel != null)
                builder.addAll(focusModel.getQuads(null, null, rand));
        }

        builder.addAll(coreModel.bake(bakery, spriteGetter, transformsFromModel, coreLoc).getQuads(null, null, rand));

        return new WandBakedModel(builder.build(), spriteGetter.apply(coreTex), Maps.immutableEnumMap(transformMap), new AttachmentOverrideHandler(bakery), true, true, this, owner, modelTransform, tfs);
    }

    @Override
    public Collection<RenderMaterial> getTextures(IModelConfiguration owner, Function<ResourceLocation, IUnbakedModel> modelGetter, Set<Pair<String, String>> missingTextureErrors) {
        return Collections.emptyList();
    }

    protected static final class AttachmentOverrideHandler extends ItemOverrideList {
        private final ModelBakery bakery;

        public AttachmentOverrideHandler(ModelBakery bakery) {
            this.bakery = bakery;
        }


        @Nullable
        @Override
        public IBakedModel resolve(@Nonnull IBakedModel originalModel, @Nonnull ItemStack stack, @Nullable ClientWorld world, @Nullable LivingEntity entity) {
            // get cap
            CapItem cap = WandItem.getCap(stack);
            // get material
            CoreItem core = WandItem.getCore(stack);
            // get variant (staff/scepter/wand)
            // TODO: improve this slightly
            ResourceLocation variant = arcLoc("wand");
            if(stack.getTag() != null)
                variant = arcLoc(stack.getOrCreateTag().getString("variant"));


            // get focus
            // nbt context comes from the focusData tag
            //FocusItem focus = WandItem.getFocus(stack);
            //CompoundNBT focusData =
            ResourceLocation capTex = cap!=null ? cap.id : arcLoc("void_cap");
            ResourceLocation coreTex = core!=null ? core.id : arcLoc("ice_wand_core");

            return new WandModelGeometry(
                    withPath(capTex, "models/wands/caps/"),
                    withPath(coreTex,"models/wands/cores/"),
                    variant,
                    arcLoc("item/wands/foci/wand_focus"))
                    .bake(
                            ((WandBakedModel) originalModel).owner,
                            bakery,
                            ModelLoader.defaultTextureGetter(),
                            ((WandBakedModel) originalModel).modelTransform,
                            originalModel.getOverrides(),
                            new ResourceLocation("arcana:does_this_do_anything_lol"));
        }
    }
}

File: src/main/kotlin/arcana/client/model/wand/WandModelGeometry.kt
package arcana.client.model.wand
import arcana.client.ClientPaths.CAPS_3D
import arcana.client.ClientPaths.CORES_3D
import arcana.common.items.ModItems
import arcana.common.items.wand.WandItem
import arcana.utils.Util.arcLoc
import arcana.utils.Util.withPath
import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import com.google.common.collect.Maps
import com.mojang.datafixers.util.Either
import com.mojang.datafixers.util.Pair
import net.minecraft.client.renderer.model.*
import net.minecraft.client.renderer.model.ItemCameraTransforms.TransformType
import net.minecraft.client.renderer.texture.AtlasTexture
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.client.world.ClientWorld
import net.minecraft.entity.LivingEntity
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.vector.TransformationMatrix
import net.minecraftforge.client.model.IModelConfiguration
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.client.model.ModelTransformComposition
import net.minecraftforge.client.model.PerspectiveMapWrapper
import net.minecraftforge.client.model.geometry.IModelGeometry
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.util.*
import java.util.function.Function


class WandModelGeometry(
    val cap: ResourceLocation,
    val core: ResourceLocation,
    val variant: ResourceLocation
) : IModelGeometry<WandModelGeometry> {

    constructor() : this(ModItems.VOID_CAP.id, ModItems.ICE_CORE.id, arcLoc("wand"))

    var LOGGER: Logger = LogManager.getLogger()

    override fun bake(owner: IModelConfiguration, bakery: ModelBakery,
        spriteGetter: Function<RenderMaterial, TextureAtlasSprite>,
        modelTransform: IModelTransform, overrides: ItemOverrideList, modelLocation: ResourceLocation
    ): IBakedModel {

        val coreTex = RenderMaterial(AtlasTexture.LOCATION_BLOCKS, core)
        val capTex = RenderMaterial(AtlasTexture.LOCATION_BLOCKS, cap)

        val variantLoc = variant.withPath{"item/wands/variants/" + it }
        val variantModel = bakery.getModel(variantLoc)
        var tfs = ItemCameraTransforms.NO_TRANSFORMS
        if (variantModel is BlockModel) {
            variantModel.textureMap["core"] = Either.left(coreTex)
            variantModel.textureMap["cap"] = Either.left(capTex)
            tfs = variantModel.transforms
        } else
            LOGGER.error("Wand model isn't a block model!")

        val builder = ImmutableList.builder<BakedQuad>()
        val rand = Random()
        val focusModel = bakery.getBakedModel(variant.withPath { "item/wands/foci/${it}_focus" }, modelTransform, spriteGetter)
        if (focusModel != null)
            builder.addAll(focusModel.getQuads(null, null, rand))

        val transformsFromModel = owner.combinedTransform
        val transformMap: ImmutableMap<TransformType, TransformationMatrix> =
            PerspectiveMapWrapper.getTransforms(ModelTransformComposition(transformsFromModel, modelTransform))
        builder.addAll(variantModel.bake(bakery, spriteGetter, transformsFromModel, variantLoc)!!.getQuads(null, null, rand))

        return WandBakedModel(builder.build(), spriteGetter.apply(coreTex), Maps.immutableEnumMap(transformMap),
                              AttachmentOverrideHandler(bakery), true, true, this,
                              owner, modelTransform, tfs)
    }

    override fun getTextures(owner: IModelConfiguration, modelGetter: Function<ResourceLocation, IUnbakedModel>, missingTextureErrors: MutableSet<Pair<String, String>>): MutableCollection<RenderMaterial> {
        return Collections.emptyList()
    }

    protected class AttachmentOverrideHandler(private val bakery: ModelBakery) : ItemOverrideList() {
        override fun resolve(originalModel: IBakedModel, stack: ItemStack, world: ClientWorld?, entity: LivingEntity?): IBakedModel? {
            // get cap
            val cap = WandItem.getCap(stack)
            // get material
            val core = WandItem.getCore(stack)
            // get variant (staff/scepter/wand)
            // TODO: improve this slightly
            var variant: ResourceLocation = arcLoc("wand")
            if (stack.tag != null)
                variant = arcLoc(stack.orCreateTag.getString("variant"))


            val capTex = cap.id
            val coreTex = core.id
            return WandModelGeometry(
                capTex.withPath { CAPS_3D + it },
                coreTex.withPath { CORES_3D + it },
                variant)
                .bake(
                    (originalModel as WandBakedModel).owner,
                    bakery,
                    ModelLoader.defaultTextureGetter(),
                    originalModel.modelTransform,
                    originalModel.getOverrides(),
                    ResourceLocation("arcana:does_this_do_anything_lol"))
        }
    }
}